checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java

```diff
@@ -201,6 +209,93 @@ private void checkOwningOverrides(
+   *
+   * <p>The return type is checked if it is annotated with {@code @PolyMustCall} because the Must
+   * Call Checker treats {@code @MustCallAlias} as an alias of {@code @PolyMustCall}.
+   *
+   * @param tree the method declaration.
+   * @param mcAtf the MustCallAnnotatedTypeFactory.
+   */
+  private void checkMustCallAliasAnnotationForMethod(
+      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
+
+    Element paramWithMustCallAliasAnno = getParameterWithMustCallAliasAnno(tree);
+    boolean isMustCallAliasAnnoOnParameter = paramWithMustCallAliasAnno != null;
+
+    if (TreeUtils.isVoidReturn(tree) && isMustCallAliasAnnoOnParameter) {
+      checker.reportWarning(
+          tree, "mustcallalias.method.return.and.param", "this method has a void return");
+      return;
+    }
+
+    AnnotatedTypeMirror returnType = mcAtf.getMethodReturnType(tree);
+    boolean isMustCallAliasAnnoOnReturnType = returnType.hasPrimaryAnnotation(PolyMustCall.class);
+
+    if (isMustCallAliasAnnoOnParameter != isMustCallAliasAnnoOnReturnType) {
+      String message =
+          isMustCallAliasAnnoOnParameter
+              ? String.format(
+                  "there is no @MustCallAlias annotation on the return type, even though parameter %s is annotated with @MustCallAlias",
+                  paramWithMustCallAliasAnno)
+              : "no parameter has a @MustCallAlias annotation, even though the return type is annotated with @MustCallAlias";
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param", message);
+    }
+  }
+
+  /**
+   * Given a constructor, a {@code @MustCallAlias} must appear in both the list of parameters and as
+   * an annotation on the constructor itself, if it is to appear at all.
+   *
+   * <p>That is, a {@code @MustCallAlias} annotation must appear on both the constructor and its
+   * parameter list, or not at all.
+   *
+   * @param tree the constructor.
+   */
+  private void checkMustCallAliasAnnotationForConstructor(MethodTree tree) {
+    ExecutableElement constructorDecl = TreeUtils.elementFromDeclaration(tree);
+    boolean isMustCallAliasAnnoOnConstructor =
+        constructorDecl != null && rlTypeFactory.hasMustCallAlias(constructorDecl);
+    Element paramWithMustCallAliasAnno = getParameterWithMustCallAliasAnno(tree);
+    boolean isMustCallAliasAnnoOnParameter = paramWithMustCallAliasAnno != null;
+
+    if (isMustCallAliasAnnoOnConstructor != isMustCallAliasAnnoOnParameter) {
+      String message =
+          isMustCallAliasAnnoOnParameter
+              ? String.format(
+                  "there is no @MustCallAlias annotation on the return type, even though parameter %s is annotated with @MustCallAlias",
+                  paramWithMustCallAliasAnno)
+              : "no parameter has a @MustCallAlias annotation, even though the return type is annotated with @MustCallAlias";
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param", message);
+    }
```

Comment:
I don't particularly like this duplicated code (same code in `checkMustCallAliasAnnotationForMethod`) but I'm not sure if extracting it into its own method would be overkill here

Reply:
I think it's fine to factor it out. I almost always prefer helper methods for this kind of thing.